components/ws-proxy/pkg/proxy/infoprovider.go

```diff
@@ -413,25 +432,69 @@ func (c *workspaceInfoCache) doInsert(info *WorkspaceInfo) {
 	}
 }
 
-func (c *workspaceInfoCache) Delete(workspaceID string) {
+func (c *workspaceInfoCache) Delete(workspaceID string, instanceID string) {
 	c.cond.L.Lock()
 	defer c.cond.L.Unlock()
 
-	info, present := c.infos[workspaceID]
-	if !present || info == nil {
+	infos, present := c.infos[workspaceID]
+	if !present || infos == nil {
 		return
 	}
-	delete(c.coordsByPublicPort, info.IDEPublicPort)
-	delete(c.infos, workspaceID)
+
+	// Keep only the active instances and public port(s) for the workspace id
+	activeInfos := []*WorkspaceInfo{}
+	activePublicPorts := make(map[string]interface{})
+	var instanceIDEPublicPort string
+	for _, info := range infos {
+		if info.InstanceID != instanceID {
```

Comment:
nit: `if` statements with an `else` clause should have the non-negated case, otherwise you have to mentally double-negate to understand the clause.